### PR TITLE
Fixed #28646 - fixed PostgreSQL generating indexes in certain case

### PR DESCRIPTION
**Pull request by bcail**
Thanks to Ariki for the test case, and Carlton for the fix.

### CODE REVIEW

File: django/db/backends/postgresql/schema.py [hunk 1]
```diff
@@ -297,7 +297,7 @@ def _alter_field(
         )
         # Added an index? Create any PostgreSQL-specific indexes.
         if (not (old_field.db_index or old_field.unique) and new_field.db_index) or (
-            not old_field.unique and new_field.unique
+            not old_field.unique and new_field._unique
```
Comment (felixxm) [inline django/db/backends/postgresql/schema.py:300]: I don't believe that is a proper solution, you need to consider all cases and write it down to a correct logic formula. We have to consider here two cases:
1. when `_like` indexes should be added
2. when `_like` indexes should be re-created because they were removed in `super()._alter_field()`

Comment (felixxm) [reply django/db/backends/postgresql/schema.py:300]: IMO, this is a regression in 9356f63a99957f01c14a9788617428a172a29fcb when the logic was simplified missing some cases.

**Review by felixxm — COMMENTED** [review]
@bcail Thanks :+1: However, I'm convinces this is not a proper solution.

File: django/db/backends/postgresql/schema.py (hunk 1 above)
Comment (bcail) [reply django/db/backends/postgresql/schema.py:300]: OK, thanks - I'll work on adding the missing tests.

File: tests/migrations/test_operations.py [hunk 2]
```diff
@@ -6048,6 +6050,46 @@ def test_indexes_ignore_swapped(self):
             )
 
 
+@unittest.skipUnless(connection.vendor == "postgresql", "PostgreSQL specific")
+class TestChangePrimaryKey(TestCase):
+    available_apps = ["migrations"]
+
+    def test(self):
+        """Regression test for #28646."""
```
[the remark this answers is not in the document]
Comment (bcail) [reply tests/migrations/test_operations.py]: Removed.

File: tests/migrations/test_operations.py [hunk 3]
```diff
@@ -6048,6 +6050,46 @@ def test_indexes_ignore_swapped(self):
             )
 
 
+@unittest.skipUnless(connection.vendor == "postgresql", "PostgreSQL specific")
+class TestChangePrimaryKey(TestCase):
```
[the remark this answers is not in the document]
Comment (bcail) [reply tests/migrations/test_operations.py]: Moved the test.

File: django/db/backends/postgresql/schema.py (hunk 1 above)
Comment (bcail) [reply django/db/backends/postgresql/schema.py]: @felixxm I've added tests and updated the fix.

Comment (bcail) [reply django/db/backends/postgresql/schema.py]: Another option is what was mentioned in the issue: we could just use "IF NOT EXISTS" and not try to determine whether the _like index is already there or not.

Comment (felixxm) [reply django/db/backends/postgresql/schema.py]: > @felixxm I've added tests and updated the fix.

We should not use `_unique` (as I said previously). Also, most of added tests seems to be redundant with the existing tests, or are redundant with each other e.g. `test_slugfields_change_primary_key` and `test_slugfields_change_primary_key_operations`.

Comment (bcail) [reply django/db/backends/postgresql/schema.py]: @felixxm What do you think of just using "IF NOT EXISTS" for creating the index? Then we wouldn't need to worry as much about all the different states of the indexes.

File: django/db/backends/postgresql/schema.py [hunk 4]
```diff
@@ -15,11 +15,11 @@ class DatabaseSchemaEditor(BaseDatabaseSchemaEditor):
     sql_delete_sequence = "DROP SEQUENCE IF EXISTS %(sequence)s CASCADE"
 
     sql_create_index = (
-        "CREATE INDEX %(name)s ON %(table)s%(using)s "
+        "CREATE INDEX IF NOT EXISTS %(name)s ON %(table)s%(using)s "
         "(%(columns)s)%(include)s%(extra)s%(condition)s"
     )
     sql_create_index_concurrently = (
-        "CREATE INDEX CONCURRENTLY %(name)s ON %(table)s%(using)s "
+        "CREATE INDEX CONCURRENTLY IF NOT EXISTS %(name)s ON %(table)s%(using)s "
```
Comment (felixxm) [inline django/db/backends/postgresql/schema.py:22]: I think it is unacceptable, because it will hide real issues from users. Users should be inform when an index already exists, as it may be a different constraint then they expect. This change is also backward incompatible as it changes the current behavior for creating indexes.

Comment (bcail) [reply django/db/backends/postgresql/schema.py:22]: OK, thanks @felixxm. I'll revert those changes.

File: django/db/backends/postgresql/schema.py [hunk 5]
```diff
@@ -274,6 +274,25 @@ def _alter_column_collation_sql(
...
+        if (
+            old_field.db_index
+            and not old_field.unique
+            and (
+                not new_field.db_index
+                or (new_field.unique and not new_field.primary_key)
+            )
+        ) or (
+            old_type != new_type and not new_field.unique and not old_field.primary_key
+        ):
```
Comment (felixxm) [inline django/db/backends/postgresql/schema.py:293]: Can you point me to where all indexes are dropped due to type change? As far as I'm aware, this should not be required to check this directly.

Comment (bcail) [reply django/db/backends/postgresql/schema.py:293]: @felixxm Here's where the _like index is dropped when changing certain types: https://github.com/django/django/blob/main/django/db/backends/postgresql/schema.py#L158.

I could move that check out into its own method, and we could call the same method when we're seeing if we need to recreate the index.

Comment (bcail) [reply django/db/backends/postgresql/schema.py]: I updated the check to match the check where the _like index is dropped.

Note: there seems to be an error on a couple of the job runners:
```
Another git process seems to be running in this repository, e.g.
an editor opened by 'git commit'. Please make sure all processes
are terminated then try again. If it still fails, a git process
may have crashed in this repository earlier:
remove the file manually to continue.
```

**Comment by sarahboyce** [comment]
buildbot, test on oracle.

File: tests/schema/tests.py [hunk 6]
```diff
@@ -5210,6 +5216,225 @@ def test_alter_field_add_db_index_to_charfield_with_unique(self):
...
+        self.assertEqual(
+            self.get_constraints_for_column(SimpleModel, "field2"),
+            [
+                "schema_simplemodel_field2_08772539_like",
+                "schema_simplemodel_field2_08772539_pk",
+            ],
+        )
+
+    @unittest.skipUnless(connection.vendor == "postgresql", "PostgreSQL specific")
+    def test_slugfields_change_primary_key_operations(self):
```
Comment (sarahboyce) [inline tests/schema/tests.py:5404]: This test is problematic (if I revert the changes, this errors and then causes ~50 tests to error) 
I believe it's because the table is not cleaned up properly but I also believe it might be already tested through `test_slugfields_change_primary_key` and can be removed

File: django/db/backends/base/schema.py [hunk 7]
```diff
@@ -951,9 +951,7 @@ def _alter_field(
                 fks_dropped.add((old_field.column,))
                 self.execute(self._delete_fk_sql(model, fk_name))
         # Has unique been removed?
-        if old_field.unique and (
-            not new_field.unique or self._field_became_primary_key(old_field, new_field)
-        ):
+        if old_field._unique and not new_field._unique:
```
Comment (sarahboyce) [inline django/db/backends/base/schema.py:954]: We're still using `_unique` here

**Review by sarahboyce — COMMENTED** [review]
Thank you for this @bcail 
Looks like there's still a few comments from @felixxm's previous review which need to be addressed here :+1:

File: tests/schema/tests.py (hunk 6 above)
Comment (bcail) [reply tests/schema/tests.py:5404]: I'm not confident that this is fully covered by other tests, but I can remove it if you want.

File: django/db/backends/base/schema.py (hunk 7 above)
Comment (bcail) [reply django/db/backends/base/schema.py:954]: I'm not clear on why we're not allowed to use the `_unique` property. Is it because it's supposed to be private on the `Field` class, so no one else is supposed to use it? If so, is there a public property that means the same thing, that we can use?

Comment (sarahboyce) [reply django/db/backends/base/schema.py:954]: Something like `if old_field.unique and not old_field.primary_key and not new_field.unique:` might work 🤔

Comment (bcail) [reply django/db/backends/base/schema.py]: OK, I used that.

File: django/db/backends/postgresql/schema.py [hunk 8]
```diff
@@ -255,6 +255,30 @@ def _alter_column_type_sql(
                 model, old_field, new_field, new_type, old_collation, new_collation
             )
 
+    def _need_to_add_new_index(self, old_field, new_field):
```
[the remark this answers is not in the document]
Comment (bcail) [reply django/db/backends/postgresql/schema.py]: I renamed the methods.

File: django/db/backends/postgresql/schema.py [hunk 9]
```diff
@@ -255,6 +255,30 @@ def _alter_column_type_sql(
...
+                not new_field.db_index
+                or (new_field.unique and not new_field.primary_key)
+            )
+        ) or (
+            (old_field.db_index or old_field.unique)
+            and (
+                (old_type.startswith("varchar") and not new_type.startswith("varchar"))
+                or (old_type.startswith("text") and not new_type.startswith("text"))
+                or (old_type.startswith("citext") and not new_type.startswith("citext"))
+            )
```
[the remark this answers is not in the document]
Comment (bcail) [reply django/db/backends/postgresql/schema.py]: I pulled out a new method for this.

File: tests/schema/tests.py (hunk 6 above)
Comment (bcail) [reply tests/schema/tests.py]: I moved the test into its own class in operations tests, and when I made it fail, it only showed 3 failures for that test class.

**Review by sarahboyce — COMMENTED** [review]
Thank you for the updates @bcail - I think we're really close ⭐

File: django/db/backends/postgresql/schema.py [hunk 10]
```diff
@@ -255,6 +262,28 @@ def _alter_column_type_sql(
...
+    def _deleted_index_should_be_recreated(
+        self, old_field, new_field, old_type, new_type
+    ):
+        if (
+            old_field.db_index
+            and not old_field.unique
+            and (
+                not new_field.db_index
+                or (new_field.unique and not new_field.primary_key)
+            )
```
[the remark this answers is not in the document]
Comment (bcail) [reply django/db/backends/postgresql/schema.py]: Thanks, I added a for a `db_index` being removed from a field, and removed the `old_field.db_index` check. I think it may have come from [https://github.com/django/django/blob/main/django/db/backends/base/schema.py#L1011](https://github.com/django/django/blob/main/django/db/backends/base/schema.py#L1011), but may not be needed here.

**Review by sarahboyce — APPROVED** [review]
Thank you for all your work on this @bcail ⭐

**Comment by sarahboyce** [comment]
Merged in 9cf9c796be8dd53bc3b11355ff39d65c81d7be6d :+1:
